Sort order

MasterMindNET opened the pull request:
From this point, you are able to set the gallery order in your "Space Menu" for space only (for now, if more needed it can be done).
You can change it from space=>modules=>configure or directly from your gallery gear => settings

Validation rules:
check if integer;
if it is an even number to the number 100;
a value greater or equal to SORT_MIN;
a value less or equal to SORT_MAX.

luke- commented on config.php:
```diff
@@ -10,9 +11,10 @@
-        ['class' => Sidebar::className(),'event' =>  BaseMenu::EVENT_INIT, 'callback' => ['humhub\modules\gallery\Events','onSpaceSidebarInit']]
+        ['class' => Menu::class,'event' => Menu::EVENT_INIT, 'callback' => [Events::class, 'onSpaceMenuInit']],
+        ['class' => ProfileMenu::class,'event' => ProfileMenu::EVENT_INIT, 'callback' => [Events::class,'onProfileMenuInit']],
+//        ['class' => SpaceSidebar::class,'event' =>  BaseMenu::EVENT_INIT, 'callback' => [Events::class,'onSpaceSidebarInit']],
```
Please remove such dead code

MasterMindNET replied:
All right! I will do it! I will remove the line number 16.

Thanks!

luke- commented on models/forms/ContainerSettings.php:
```diff
@@ -62,18 +71,40 @@ public function rules()
             [['snippetGallery', 'hideSnippet'], 'integer'],
-            [['snippetGallery'], 'containerGallery']
+            [['snippetGallery'], 'containerGallery'],
+            [['sortOrder'], 'validateSortOrder'], //'in', 'range' => [0,100], 'message'=>'Please enter a value from 0 to 300 in increments of 100'
```
Why not use Yii2 build in number validator?
https://www.yiiframework.com/doc/api/2.0/yii-validators-numbervalidator

```
[['sortOrder'], 'number', 'min' => static::SORT_MIN, 'max' => static::SORT_MAX]
```

MasterMindNET replied:
Good day,@luke-!

I have created a custom validation to have more advanced validation and output messages.
It checks whether the value is in the range from MIN and MAX and with a step of 100 only. So the values that are only acceptable is this kind of list [0, 100, 200...5000] 

if you want to have no restrictions as you described, we can make it as you have mentioned above.

Thank you!

luke- commented on views/setting/index.php:
```diff
@@ -22,6 +22,8 @@
             <?php $form = ActiveForm::begin() ?>
                 <?= $form->field($settings, 'snippetGallery')->dropDownList($settings->getGallerySelection())?>
+                <?= $form->field($settings, 'sortOrder')->textInput(['type' => 'number']) ?>
+<!--        'min' =>  $settings::SORT_MIN, 'max' => $settings::SORT_MAX, 'placeholder' => Yii::t('GalleryModule.base', 'from ' . $settings::SORT_MIN . ' to ' . $settings::SORT_MAX . ' in increments of 100'), 'maxLength' => 4]-->
```
Remove

MasterMindNET replied:
Do you mean "Remove" the commented out line number 26, isn't it?

Thanks!

luke- replied:
Yes, just remove Line 26 which is commented out anyway.

MasterMindNET replied:
I have already pushed a few commits (I believe you can see it in my old PR).
I leave it like this:
`<?= $form->field($settings, 'sortOrder')->textInput(['type' => 'number', 'placeholder' => Yii::t('GalleryModule.base', 'from ' . $settings::SORT_MIN . ' to ' . $settings::SORT_MAX . ' in increments of 100'),]) ?>`

Please, let me know if it what you wanted.
Thank you for your quick reply!

luke- commented:
Thanks for your great work so far!

Following remarks:

1.) The sort order is mainly about the position of the snippet (blue box) in the sidebar (red box). Screenshot below:
![image](https://user-images.githubusercontent.com/4736168/85058250-1b5b2d80-b1a2-11ea-86ba-d6327eb4d5cb.png)
If it's unclear how to set the sort order, please let me know.


2.) In my opinion we do not need a deeper check of the sort order input field. It is suffient to check it's a number. (0-32000).
This simple approach is currently being used in other modules as well. Maybe we should find a more elegant solution here in future for all modules.


3.) Maybe we should add the term `snippet` to the `sortOrder` vars (settings & model) to make sure there is no confusion. e.g. `snippetSortOrder`.

MasterMindNET commented:
Good day, @luke-!

How can I add the gallery to the right sidebar?
> 1.) The sort order is mainly about the position of the snippet (blue box) in the sidebar (red box). Screenshot below:
> ![image](https://user-images.githubusercontent.com/4736168/85058250-1b5b2d80-b1a2-11ea-86ba-d6327eb4d5cb.png)
> If it's unclear how to set the sort order, please let me know.

Fixed:
> 2.) In my opinion we do not need a deeper check of the sort order input field. It is suffient to check it's a number. (0-32000).
> This simple approach is currently being used in other modules as well. Maybe we should find a more elegant solution here in future for all modules.

I have changed the name of this property.
> 3.) Maybe we should add the term `snippet` to the `sortOrder` vars (settings & model) to make sure there is no confusion. e.g. `snippetSortOrder`.

luke- commented:
@MasterMindNET 

1.) The gallery sidebar snippet should be automatically added on the Space "Stream" page if images exists.
3.) Yes, both options the "Choose snippet gallery" (1) and the new "Sort Order" input are both related to the snippet.

luke- commented:
@MasterMindNET 

PR is looking good so far. The only thing left now is to change the sort order.

- Remove lines 40-42
https://github.com/humhub-contrib/gallery/blob/3c1d0ebda7bbf0101ee14b167ecbd040ad723ffb/Events.php#L40-L42 
- Add the sort order here. Replace 0 with the current sort order
https://github.com/humhub-contrib/gallery/blob/3c1d0ebda7bbf0101ee14b167ecbd040ad723ffb/Events.php#L70 

e.g. 
``` 
(int) $module->settings->contentContainer($event->sender->space)->get(ContainerSettings::SETTING_SORT_ORDER)
```

MasterMindNET commented:
> @MasterMindNET
> 
> 1.) The gallery sidebar snippet should be automatically added on the Space "Stream" page if images exists.
> 3.) Yes, both options the "Choose snippet gallery" (1) and the new "Sort Order" input are both related to the snippet.

I do not have the gallery snippet in my space. Where can I look at to find the reason why it is now displaying to me?

Each gallery has at least one image, but there is no any of them on my stream page.
Here is my setup:
![image](https://user-images.githubusercontent.com/22347676/85223777-1f6b9300-b3ce-11ea-9a06-bfb4257b16ab.png)

luke- commented on models/forms/ContainerSettings.php:
```diff
@@ -84,7 +94,8 @@ public function attributeLabels()
             'snippetGallery' => Yii::t('GalleryModule.base', 'Choose snippet gallery'),
-            'hideSnippet' => Yii::t('GalleryModule.base', 'Don\'t show the gallery snippet in this space.')
+            'hideSnippet' => Yii::t('GalleryModule.base', 'Don\'t show the gallery snippet in this space.'),
+            'snippetSortOrder' => Yii::t('GalleryModule.config', 'Sort order'),
```
Please use `GalleryModule.base` as Translation Message category

luke- commented on views/setting/index.php:
```diff
@@ -22,6 +22,7 @@
         <?php if($settings->hasGallery()) : ?>
             <?php $form = ActiveForm::begin() ?>
                 <?= $form->field($settings, 'snippetGallery')->dropDownList($settings->getGallerySelection())?>
+                <?= $form->field($settings, 'snippetSortOrder')->textInput(['type' => 'number', 'placeholder' => Yii::t('GalleryModule.base', 'from ' . $settings::SORT_MIN . ' to ' . $settings::SORT_MAX . ' in increments of 100'),]) ?>
```
Please remove placeholder text here.We need to find here a solution in future.

luke- commented on widgets/views/galleryListEntryAdd.php:
```diff
@@ -20,7 +20,7 @@
         <div class="panel-body">
             <a class="backgroundSuccess" <?= yii\bootstrap\Html::renderTagAttributes($htmlOptions) ?> href="<?= $addActionUrl ?>">
-                <img class="padding15perc" src="<?= Yii::$app->getModule('gallery')->getAssetsUrl() . '/plus.svg' ?>" />
+                <img class="padding15perc" src="<?= Yii::$app->getModule('gallery')->getAssetsUrl() . '/plus.svg' ?>" alt="add gallery image" />
```
Can you please use `Yii::t here?`

MasterMindNET replied:
Good day, @luke-!

Oh, sure. I have missed this one. Sorry about that.
I added this (and some more), please take a look at the last commits.

Thank you!

luke- commented:
@MasterMindNET Hmm, I'm not sure why the gallery  snippet not appears on your installation.

![image](https://user-images.githubusercontent.com/4736168/85228916-c2c8a200-b3e6-11ea-8273-416259123fbd.png)

Make sure you have selected a gallery with some images and the "Don't show the gallery snippet" is not active.

MasterMindNET commented:
> @MasterMindNET Hmm, I'm not sure why the gallery snippet not appears on your installation.
> 
> ![image](https://user-images.githubusercontent.com/4736168/85228916-c2c8a200-b3e6-11ea-8273-416259123fbd.png)
> 
> Make sure you have selected a gallery with some images and the "Don't show the gallery snippet" is not active.

Thanks!
That is what I needed.
